### ISSUE TITLE
Timeout error for call into Lua sandbox

### ISSUE THREAD

**Issue by mhball** [erlio/vernemq#798]
### Environment

VerneMQ Version: 1.3.2
OS: Ubuntu 16.04.4
Erlang/OTP version (if building from source): N/A

### Expected behavior

- We have a Lua script running authentication on publish.  It does a simple string comparison on the username and a regex match on the topic. 

The function is as follows:
```
function andium_auth_on_publish(pub)
    if pub.username ~= READ_ONLY and (pub.username==CLOUD_USER or string.match(pub.topic, "^[^/]+/[^/]+/([^/]+)$") == pub.username) then
        return true
    end

    return false
end
```
With `READ_ONLY` and `CLOUD_USER` being static variables set at the top of the script.

### Actual behaviour

 - It's been working well for some time; however, we've now seen the following error, which caused a message to be dropped:

```
2018-08-12 18:26:12.518 [error] <0.17602.9>@vmq_diversity_script:call_function:64 can't call into Lua sandbox for function auth_on_publish due to {timeout,{gen_server,call,[<0.350.0>,{call_function,auth_on_publish,[{username,<<"urn:andium:nid:flow-rkxu:00313753-3431-3435-3134-303436303130">>},{mountpoint,<<>>},{client_id,<<"urn:andium:nid:flow-rkxu:00313753-3431-3435-3134-303436303130">>},{qos,1},{topic,<<"urn:andium:cid:e2805630-3760-437b-bcd0-1985dc33d67d/RKXU.MQTTstream.Test/urn:andium:nid:flow-rkxu:00313753-3431-3435-3134-303436303130">>},{payload,<<...json payload removed...>>},{retain,false}]}]}}
```
I can't see how the function itself would timeout - there's no external call.  Is it possible the timeout is in the Lua function handling in the broker?

**Comment by ioolkos** [comment]
@mhball thanks for the issue. Was this happening under increased load? (more clients, messages)?

Lua experts, please step up! :) (@dergraf)

**Comment by mhball** [comment]
@ioolkos No, the load has been stable. Less than 10 clients sending a message every 4 seconds and less than 5 clients subscribing.

In case it affects anything, we do also use a Lua plugin to perform auth_on_subscribe (similar topic regex check), and perform auth_on_register (does make a HTTP request).

**Comment by mhball** [comment]
We saw this same error again today, and unfortunately followed by some more serious errors:
```
2018-08-16 00:21:05.175 [error] <0.348.0>@vmq_diversity_http:request:57 http request failure for get <<"https://auth.prod.andium.net/authentication/v1/system
?applicationId=mqtt">> due to closed
2018-08-16 00:21:05.175 [error] <0.348.0>@vmq_diversity_script_state:handle_info:171 can't call function auth_on_register with args [{addr,<<"10.61.57.210">>
},{port,46272},{mountpoint,<<>>},{client_id,<<"mqttjs_0eaf1377">>},{username,<<"urn:andium:eid:123456-123456-123456">>},{password,<<"...removed...">>},{clea
n_session,true}] in "/etc/vernemq/andium.lua" due to {illegal_index,false,<<"ref">>}
```

We saw the above repeated a few times.  Followed by

```
2018-08-16 00:38:48.202 [error] <0.348.0> gen_server <0.348.0> terminated with reason: no function clause matching vmq_diversity_script_state:handle_info({ss
l_closed,{sslsocket,{gen_tcp,#Port<0.106505>,tls_connection,undefined},<0.8752.13>}}, {state,1,{luerl,{array,26,100,undefined,{{{table,{array,0,10,nil,10},{{
{{{empty,<<"CLOUD_PAS...">>,...},...},...},...},...},...},...},...}},...},...}) line 162
2018-08-16 00:38:48.203 [error] <0.348.0> CRASH REPORT Process <0.348.0> with 0 neighbours exited with reason: no function clause matching vmq_diversity_scri
pt_state:handle_info({ssl_closed,{sslsocket,{gen_tcp,#Port<0.106505>,tls_connection,undefined},<0.8752.13>}}, {state,1,{luerl,{array,26,100,undefined,{{{tabl
e,{array,0,10,nil,10},{{{{{empty,<<"CLOUD_PAS...">>,...},...},...},...},...},...},...},...}},...},...}) line 162 in gen_server:terminate/7 line 812
2018-08-16 00:38:48.203 [error] <0.347.0> Supervisor {<0.347.0>,vmq_diversity_script_sup_sup} had child {vmq_diversity_script_state,1} started with vmq_diver
sity_script_state:start_link(1, "/etc/vernemq/andium.lua") at <0.348.0> exit with reason no function clause matching vmq_diversity_script_state:handle_info({
ssl_closed,{sslsocket,{gen_tcp,#Port<0.106505>,tls_connection,undefined},<0.8752.13>}}, {state,1,{luerl,{array,26,100,undefined,{{{table,{array,0,10,nil,10},
{{{{{empty,<<"CLOUD_PAS...">>,...},...},...},...},...},...},...},...}},...},...}) line 162 in context child_terminated
```

And then all auth stopped and we saw the same missing hook error for all:

```
2018-08-16 00:38:50.208 [error] <0.8765.13>@vmq_mqtt_fsm:check_user:538 can't authenticate client {[],<<"urn:andium:nid:662f4469-90da-4be6-9158-6013c0bfd041"
>>} due to
                                no_matching_hook_found
```

Any help to debug would be appreciated.  Unfortunately this caused us to drop production data.

**Comment by ioolkos** [comment]
@dergraf @larshesel could you look into this? we have seen somewhat similar issues where the reason was an error in the Lua script. 
This seems to be a working script though, that then degrades.

**Comment by larshesel** [comment]
I haven't looked too closely at this, but one thing jumped out: 

```
2018-08-16 00:21:05.175 [error] <0.348.0>@vmq_diversity_script_state:handle_info:171 can't call function auth_on_register with args [{addr,<<"10.61.57.210">>
},{port,46272},{mountpoint,<<>>},{client_id,<<"mqttjs_0eaf1377">>},{username,<<"urn:andium:eid:123456-123456-123456">>},{password,<<"...removed...">>},{clea
n_session,true}] in "/etc/vernemq/andium.lua" due to {illegal_index,false,<<"ref">>}
```

The error `{illegal_index,false,<<"ref">>}` usually means there is a bug/runtime exception in the lua script. Could it be that an otherwise unused path through the code was taken which caused this error? Or was the `/etc/vernemq/andium.lua` script changed recently which could have introduced a bug?

I think this looks to me like too many errors restarted a supervisor but for some reason the lua script was no longer loaded after the supervisor restart. Does things go back to a functioning state if you stop/start the `vmq_diversity` plugin?

**Comment by larshesel** [comment]
Could it be you're running into this: https://github.com/erlio/vernemq/issues/588 (tldr: one always has to call `http.body` otherwise the connection won't be released to the pool)

**Comment by larshesel** [comment]
Just created #802 to address this issue once and (hopefully) for all - of course - not sure if that's the issue you're running into.

**Comment by dergraf** [comment]
@mhball What is the status of the issue. Our current assumption is that this has been solved in the `1.5.0`  version.
